discojs/src/processing/index.ts

```diff
@@ -0,0 +1,172 @@
+        processing.normalize(
+          processing.removeAlpha(processing.resize(IMAGE_W, IMAGE_H, image)),
+        ),
+        processing.indexInList(label, LABEL_LIST),
+      ]) as Dataset<ModelEncoded[D]>;
+    }
+    case "tabular": {
+      // cast as typescript doesn't reduce generic type
+      const d = dataset as Dataset<Raw["tabular"]>;
+      const { inputColumns, outputColumn } =
+        task.trainingInformation as TrainingInformation<"tabular">;
+
+      return d.map((row) => {
+        const output = processing.extractColumn(row, outputColumn);
+
+        return [
+          extractToNumbers(inputColumns, row),
+          // TODO sanitization doesn't care about column distribution
+          output !== "" ? processing.convertToNumber(output) : 0,
+        ];
+      }) as Dataset<ModelEncoded[D]>;
+    }
+    case "text": {
+      // cast as typescript doesn't reduce generic type
+      const d = dataset as Dataset<Raw["text"]>;
+      const t = task as Task<"text">;
+
+      const tokenizer = await models.getTaskTokenizer(t);
+      const maxLength =
+        t.trainingInformation.maxSequenceLength ??
+        (tokenizer.model_max_length as number) + 1;
+
+      return d
+        .map((line) =>
+          processing.tokenizeAndLeftPad(line, tokenizer, maxLength),
+        )
+        .map((tokens) => [tokens.pop(), tokens.last()]) as Dataset<
+        ModelEncoded[D]
+      >;
+    }
+  }
+}
+
+export async function preprocessWithoutLabel<D extends DataType>(
+  task: Task<D>,
+  dataset: Dataset<RawWithoutLabel[D]>,
+): Promise<Dataset<ModelEncoded[D][0]>> {
+  switch (task.trainingInformation.dataType) {
+    case "image": {
+      // cast as typescript doesn't reduce generic type
+      const d = dataset as Dataset<RawWithoutLabel["image"]>;
+      const { IMAGE_H, IMAGE_W } =
+        task.trainingInformation as TrainingInformation<"image">;
+
+      return d.map((image) =>
+        processing.normalize(
+          processing.removeAlpha(processing.resize(IMAGE_W, IMAGE_H, image)),
+        ),
+      );
+    }
+    case "tabular": {
+      // cast as typescript doesn't reduce generic type
+      const d = dataset as Dataset<Raw["tabular"]>;
+      const { inputColumns } =
+        task.trainingInformation as TrainingInformation<"tabular">;
+
+      return d.map((row) => extractToNumbers(inputColumns, row));
+    }
+    case "text": {
+      // cast as typescript doesn't reduce generic type
+      const d = dataset as Dataset<Raw["text"]>;
+      const t = task as Task<"text">;
+
+      const tokenizer = await models.getTaskTokenizer(t);
+      const maxLength =
+        t.trainingInformation.maxSequenceLength ??
+        (tokenizer.model_max_length as number) + 1;
+
+      return d
+        .map((line) =>
+          processing.tokenizeAndLeftPad(line, tokenizer, maxLength),
+        )
+        .map((tokens) => tokens.pop());
+    }
+  }
+}
+
+export async function postprocess<D extends DataType>(
```

Comment:
hum, it seems that I'm not even using `postprocess`; well, I update the validator to now use it, and to output inferred values rather than numbers.
also added a few lines to `types.Inferred` to explain more what each values should be